File: agents/github-action-agent/src/tonic_agent/github_comments.py
"""Deterministic PR / review comment bodies with hidden idempotency markers."""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass
from enum import Enum

from .comment_chunking import continuity_footer, markdown_code_fences, markdown_text_fences


class CommentMode(str, Enum):
    SUMMARY_ONLY = "summary-only"
    FILE_INLINE = "file+inline"
    INLINE_ONLY = "inline-only"
    ALL = "all"


class Verbosity(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


def marker_summary(run_id: str) -> str:
    return f"<!-- tonic-agent:summary:{run_id} -->"


def summary_upsert_prefix() -> str:
    """Stable substring used to find an existing Tonic summary comment to update."""
    return "<!-- tonic-agent:summary"


def marker_file(path: str, content_hash: str) -> str:
    return f"<!-- tonic-agent:file:{path}:{content_hash} -->"


def file_upsert_prefix(path: str) -> str:
    return f"<!-- tonic-agent:file:{path}:"


def marker_inline(path: str, start: int, end: int, kind: str) -> str:
    return f"<!-- tonic-agent:inline:{path}:{start}:{end}:{kind} -->"


def marker_orphan(
    path: str, start: int, end: int, kind: str, reason: str
) -> str:
    return f"<!-- tonic-agent:orphan-inline:{path}:{start}:{end}:{kind}:{reason} -->"


def _digest(*parts: str) -> str:
    h = hashlib.sha256()
    for p in parts:
        h.update(p.encode("utf-8", errors="replace"))
        h.update(b"|")
    return h.hexdigest()[:16]


def build_summary_body(
    run_id: str,
    pr_title: str,
    files: list[dict],
    *,
    verbosity: Verbosity = Verbosity.MEDIUM,
) -> str:
    """PR-level summary comment (Markdown)."""
    lines = [
        marker_summary(run_id),
        "",
        "## Tonic merge report",
        "",
        f"**PR:** {pr_title}",
        "",
        "Per-file conflict summaries (Tonic markers) are listed below when present.",
        "",
    ]
    if verbosity != Verbosity.LOW:
        lines.append(
            markdown_code_fences(
                "json",
                "Per-file metadata",
                json.dumps({"files": files}, indent=2),
            )
        )
    else:
        lines.append(f"- **Files analyzed:** {len(files)}")
    lines.append("")
    return "\n".join(lines)


@dataclass
class FileReviewComment:
    path: str
    body: str
    line: int | None
    side: str | None


def build_file_top_comment(
    path: str,
    left_lines: list[str],
    right_lines: list[str],
    annotated: list[str],
    *,
    ai_note: str | None = None,
    marker_branch: str | None = None,
    merge_report_artifact: str | None = None,
) -> str:
    """Single issue comment for a file path (non-inline)."""
    digest = _digest(path, "\n".join(annotated))
    body = markdown_text_fences(
        "Tonic annotated merge output (left = base, right = head)",
        "\n".join(annotated),
    )
    lines = [
        marker_file(path, digest),
        "",
        f"### `{path}`",
        "",
        body,
        "",
        f"_Left lines: {len(left_lines)} · Right lines: {len(right_lines)}_",
        continuity_footer(
            marker_branch=marker_branch,
            merge_report_artifact=merge_report_artifact,
        ),
    ]
    if ai_note:
        lines.extend(
            [
                "",
                "---",
                "",
                "**AI suggestion (optional):**",
                "",
                markdown_text_fences("AI output", ai_note),
            ]
        )
    return "\n".join(lines)


def build_orphan_inline_thread_comment(
    path: str,
    start_line: int,
    end_line: int,
    conflict_kind: str,
    reason: str,
    snippet: str,
    *,
    extra_sections: str | None = None,
    marker_branch: str | None = None,
    merge_report_artifact: str | None = None,
) -> str:
    hint = (
        "Head line span is **ambiguous** (duplicate hunks). Anchored at line 1 for thread visibility."
        if reason == "ambiguous"
        else "Head line span is **unmapped** (empty side or no match). Anchored at line 1 for thread visibility."
    )
    fence = markdown_text_fences("Tonic markers for this region", snippet)
    lines = [
        marker_orphan(path, start_line, end_line, conflict_kind, reason),
        "",
        f"**Tonic orphan** `{conflict_kind}` (annotated output lines {start_line}–{end_line})",
        "",
        hint,
        "",
        fence,
        continuity_footer(
            marker_branch=marker_branch,
            merge_report_artifact=merge_report_artifact,
        ),
    ]
    if extra_sections:
        lines.extend(["", "---", "", extra_sections.rstrip(), ""])
    return "\n".join(lines).rstrip() + "\n"


def build_inline_thread_comment(
    path: str,
    start_line: int,
    end_line: int,
    conflict_kind: str,
    snippet: str,
    *,
    extra_sections: str | None = None,
    marker_branch: str | None = None,
    merge_report_artifact: str | None = None,
) -> str:
    """PR review inline comment on a diff line (body only)."""
    fence = markdown_text_fences(
        "Tonic markers and hunks for this region",
        snippet,
    )
    lines = [
        marker_inline(path, start_line, end_line, conflict_kind),
        "",
        f"**Tonic conflict** `{conflict_kind}` (annotated output lines {start_line}–{end_line})",
        "",
        fence,
        continuity_footer(
            marker_branch=marker_branch,
            merge_report_artifact=merge_report_artifact,
        ),
    ]
    if extra_sections:
        lines.extend(["", "---", "", extra_sections.rstrip(), ""])
    return "\n".join(lines).rstrip() + "\n"


def check_run_annotation_payload(
    path: str,
    message: str,
) -> dict:
    """JSON-serializable object for optional Checks API (annotations) — caller supplies conclusion."""
    return {
        "path": path,
        "annotation_level": "warning",
        "message": message[:500],
    }

File: agents/github-action-agent-node/src/githubComments.ts
import * as crypto from "node:crypto";
import {
  continuityFooter,
  markdownCodeFences,
  markdownTextFences,
  type ContinuityHints,
} from "./commentChunking";

export type CommentMode = "summary-only" | "file+inline" | "inline-only" | "all";
export type Verbosity = "low" | "medium" | "high";

export function markerSummary(runId: string): string {
  return `<!-- tonic-agent:summary:${runId} -->`;
}

export function markerFile(path: string, contentHash: string): string {
  return `<!-- tonic-agent:file:${path}:${contentHash} -->`;
}

export function markerInline(path: string, start: number, end: number, kind: string): string {
  return `<!-- tonic-agent:inline:${path}:${start}:${end}:${kind} -->`;
}

export function markerOrphan(
  path: string,
  start: number,
  end: number,
  kind: string,
  reason: "ambiguous" | "unmapped",
): string {
  return `<!-- tonic-agent:orphan-inline:${path}:${start}:${end}:${kind}:${reason} -->`;
}

export function summaryUpsertPrefix(): string {
  return "<!-- tonic-agent:summary";
}

export function fileUpsertPrefix(path: string): string {
  return `<!-- tonic-agent:file:${path}:`;
}

function digest(...parts: string[]): string {
  const h = crypto.createHash("sha256");
  for (const p of parts) {
    h.update(p, "utf8");
    h.update("|", "utf8");
  }
  return h.digest("hex").slice(0, 16);
}

export function buildSummaryBody(
  runId: string,
  prTitle: string,
  files: Array<Record<string, unknown>>,
  verbosity: Verbosity = "medium",
): string {
  const lines = [
    markerSummary(runId),
    "",
    "## Tonic merge report",
    "",
    `**PR:** ${prTitle}`,
    "",
    "Per-file conflict summaries (Tonic markers) are listed below when present.",
    "",
  ];
  if (verbosity !== "low") {
    lines.push(
      markdownCodeFences("json", "Per-file metadata", JSON.stringify({ files }, null, 2)),
    );
  } else {
    lines.push(`- **Files analyzed:** ${files.length}`);
  }
  lines.push("");
  return lines.join("\n");
}

export function buildFileTopComment(
  filePath: string,
  leftLines: string[],
  rightLines: string[],
  annotated: string[],
  continuity?: ContinuityHints | null,
): string {
  const contentHash = digest(filePath, annotated.join("\n"));
  const body = markdownTextFences(
    "Tonic annotated merge output (left = base, right = head)",
    annotated.join("\n"),
  );
  return [
    markerFile(filePath, contentHash),
    "",
    `### \`${filePath}\``,
    "",
    body,
    "",
    `_Left lines: ${leftLines.length} · Right lines: ${rightLines.length}_`,
    continuityFooter(continuity ?? undefined),
  ]
    .filter(Boolean)
    .join("\n");
}

export function buildOrphanInlineComment(
  path: string,
  startLine: number,
  endLine: number,
  conflictKind: string,
  reason: "ambiguous" | "unmapped",
  snippet: string,
  extraSections?: string | null,
  continuity?: ContinuityHints | null,
): string {
  const hint =
    reason === "ambiguous"
      ? "Head line span is **ambiguous** (duplicate hunks). Anchored at line 1 for thread visibility."
      : "Head line span is **unmapped** (empty side or no match). Anchored at line 1 for thread visibility.";
  const fence = markdownTextFences("Tonic markers for this region", snippet);
  const lines = [
    markerOrphan(path, startLine, endLine, conflictKind, reason),
    "",
    `**Tonic orphan** \`${conflictKind}\` (annotated output lines ${startLine}–${endLine})`,
    "",
    hint,
    "",
    fence,
    continuityFooter(continuity ?? undefined),
  ];
  if (extraSections?.trim()) {
    lines.push("", "---", "", extraSections.trim(), "");
  }
  return lines.join("\n").replace(/\n+$/, "\n") + "\n";
}

export function buildInlineThreadComment(
  path: string,
  startLine: number,
  endLine: number,
  conflictKind: string,
  snippet: string,
  extraSections?: string | null,
  continuity?: ContinuityHints | null,
): string {
  const fence = markdownTextFences("Tonic markers and hunks for this region", snippet);
  const lines = [
    markerInline(path, startLine, endLine, conflictKind),
    "",
    `**Tonic conflict** \`${conflictKind}\` (annotated output lines ${startLine}–${endLine})`,
    "",
    fence,
    continuityFooter(continuity ?? undefined),
  ];
  if (extraSections?.trim()) {
    lines.push("", "---", "", extraSections.trim(), "");
  }
  return lines.join("\n").replace(/\n+$/, "\n") + "\n";
}

export type { ContinuityHints };
